src/common/WeightedPriorityQueue.h

```diff
@@ -0,0 +1,359 @@
+      if (i == q.end()) {
+	return 0;
+      }
+      unsigned count = i->second.size();
+      q_size -= count;
+      if (out) {
+	for (typename ListPairs::reverse_iterator j =
+	       i->second.rbegin();
+	     j != i->second.rend();
+	     ++j) {
+	  out->push_front(j->second);
+	}
+      }
+      if (i == cur) {
+	++cur;
+      }
+      q.erase(i);
+      if (cur == q.end()) {
+	cur = q.begin();
+      }
+      return count;
+    }
+
+    void dump(ceph::Formatter *f) const {
+      f->dump_int("num_keys", q.size());
+      if (!empty()) {
+	f->dump_int("first_item_cost", front().first);
+      }
+    }
+  };
+
+  unsigned high_size, wrr_size;
+  unsigned max_cost;
+
+  typedef std::map<unsigned, SubQueue> SubQueues;
+  SubQueues high_queue;
+  SubQueues queue;
+  typename SubQueues::reverse_iterator dq;
+
+  SubQueue *create_queue(unsigned priority) {
+    typename SubQueues::iterator p = queue.find(priority);
+    if (p != queue.end()) {
+      return &p->second;
+    }
+    total_priority += priority;
+    SubQueue *sq = &queue[priority];
+    return sq;
+  }
+
+  void remove_queue(unsigned priority) {
+    assert(queue.count(priority));
+    dq = (typename SubQueues::reverse_iterator) queue.erase(queue.find(priority));
+    if (dq == queue.rend()) {
+      dq = queue.rbegin();
+    }
+    total_priority -= priority;
+    assert(total_priority >= 0);
+  }
+
+public:
+  WeightedPriorityQueue(unsigned max_per, unsigned min_c)
+    : total_priority(0),
+      high_size(0),
+      wrr_size(0),
+      max_cost(0),
+      dq(queue.rbegin())
+  {
+    srand(time(0));
+  }
+
+  unsigned length() const override final {
+    return high_size + wrr_size;
+  }
+
+  void remove_by_filter(
+      std::function<bool (T)> f, std::list<T> *removed = 0) override final {
+    for (typename SubQueues::iterator i = queue.begin();
+	 i != queue.end(); ++i) {
+      wrr_size -= i->second.remove_by_filter(f, removed);
+      unsigned priority = i->first;
+      if (i->second.empty()) {
+	remove_queue(priority);
+      }
+    }
+    for (typename SubQueues::iterator i = high_queue.begin();
+	 i != high_queue.end();
+	 ) {
+      high_size -= i->second.remove_by_filter(f, removed);
+      if (i->second.empty()) {
+	high_queue.erase(i++);
+      } else {
+	++i;
+      }
+    }
+  }
+
+  void remove_by_class(K k, std::list<T> *out = 0) override final {
+    for (typename SubQueues::iterator i = queue.begin();
+	 i != queue.end(); ++i) {
+      wrr_size -= i->second.remove_by_class(k, out);
+      unsigned priority = i->first;
+      if (i->second.empty()) {
+	remove_queue(priority);
+      }
+    }
+    for (typename SubQueues::iterator i = high_queue.begin();
+	 i != high_queue.end();
+	 ) {
+      high_size -= i->second.remove_by_class(k, out);
+      if (i->second.empty()) {
+	high_queue.erase(i++);
+      } else {
+	++i;
+      }
+    }
+  }
+
+  void enqueue_strict(K cl, unsigned priority, T item) override final {
```

Comment:
K is just an integer, so I don't think we would buy much, but T may be worthwhile pursuing. Right now, we are just looking to get the new queue in. Later, I was going to spend some more time later trying to optimize it further and look at intrusive lists which from what I understand would just deal with pointers anyways.